Incorporate feedback from reviews of GMSA KEP

ddebroy opened the pull request:
1. Introduced a CRD for tracking GMSA credspecs (instead of configmaps). 
2. Clarified goals/non-goals around restricting users within namespaces from using GMSAs.
3. Incorporated feedback around making admission controller look-up and populate cred spec data rather than CRI.

wk8 commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,152 @@ A developer has a Windows web app that they would like to deploy in a container
 
-2. A Kubernetes admin will create a configmap (in the desired namespace) and populate it with configuration information associated with the GMSA:
-  - Run a Windows PowerShell script to verify the account exists, and capture enough data to uniquely identify the GMSA into a JSON file (credspec). This doesn't contain any passwords or crypto secrets.
+2. A Kubernetes cluster admin (or cluster setup tools or an operator) will install a cluster scoped CRD (of kind GMSACredSpec) whose instances will be used to store GMSA credential spec configuration:
```
Do we want that in alpha though, instead of using a `ConfigMap`? I thought we wanted to avoid API changes at all for the alpha version?

ddebroy replied:
@wk8 using a CRD does not require core API changes. It is a way to introduce/extend Kubernetes.

wk8 replied:
Should have done my research before opening my big mouth, my bad :)

ddebroy replied:
Not at all .. good questions - please keep them coming!

wk8 commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,152 @@ A developer has a Windows web app that they would like to deploy in a container
-7. The Windows OCI implementation validates the credspec and fails to start the container if its invalid or access to the GMSA from the node is denied.
-8. The container starts with the Windows Group Managed Service Account as expected, and can authenticate to the database.
-9. During any subsequent pod update, any changes to a pod's `securityContext` is blocked (as is the case today) by `ValidatePodUpdate`.
+5. Application admins will deploy app pods that require a GMSA identity along with a Service Account authorized to use the GMSAs. There will be two options regarding how to specify the GMSACredSpec for pods and containers:
```
Why do we want both options? Would it be less confusing to users to just stick with one for now?

ddebroy replied:
I am calling out both options here for completeness. As called out towards the end of the bullet, users will typically just use the first option which is more user friendly. The second option (inline/expanded form) is an artifact of the design suggestion to keep the CRI changes simple and avoid having to look up objects in a much later stage of the pod creation process and potentially failing there along with other access control concerns. These are detailed in the section `Specifying only the name of GMSACredSpec objects in pod spec fields/annotations` under `Alternatives`

wk8 commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,152 @@ A developer has a Windows web app that they would like to deploy in a container
-9. During any subsequent pod update, any changes to a pod's `securityContext` is blocked (as is the case today) by `ValidatePodUpdate`.
+5. Application admins will deploy app pods that require a GMSA identity along with a Service Account authorized to use the GMSAs. There will be two options regarding how to specify the GMSACredSpec for pods and containers:
+
+  - Specify the name of the desired GMSACredSpec (e.g. `webapp1-credspec`): In the alpha stage of this feature, the name of the desired GMSACredSpec can be set through an annotation on the pod: `pod.alpha.kubernetes.io/windows-gmsa-credspec-name` as well as for each container through annotations of the form: `container.windows-gmsa-credspec-name.alpha.kubernetes.io/container-name`. In the beta st ...
```
I don't know if there are precedents within k8s for these annotation names, but wouldn't something like `container.alpha.kubernetes.io/windows-gmsa-credspec-name/<container-name>` be more consistent with the pod-level one (by not having the "kind" `windows-gmsa-credspec-name` in the "hostname" portion)?

wk8 replied:
I'll try with 2 `/`s first, and if not go with yours if that's okay then.

ddebroy replied:
The 2 `/`s is failing for something like `test.k8s.io/container/foo`: 
```The Deployment "nginx-deployment" is invalid: spec.template.annotations: Invalid value: "test.k8s.io/container/foo": a qualified name must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]') with an optional DNS subdomain prefix and '/' (e.g. 'example.com/MyName')```

wk8 commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,152 @@ A developer has a Windows web app that they would like to deploy in a container
+
+  - Specify the name of the desired GMSACredSpec (e.g. `webapp1-credspec`): In the alpha stage of this feature, the name of the desired GMSACredSpec can be set through an annotation on the pod: `pod.alpha.kubernetes.io/windows-gmsa-credspec-name` as well as for each container through annotations of the form: `container.windows-gmsa-credspec-name.alpha.kubernetes.io/container-name`. In the beta st ...
+
+  - Specify the contents of the `credpec` field of GMSACredSpec that gets passed down to the runtime: In the alpha stage of this feature, a JSON representation of the contents of the desired GMSACredSpec can be set through an annotation on the pod: `pod.alpha.kubernetes.io/windows-gmsa-credspec` as well as for each container through annotations of the form: `container.windows-gmsa-credspec.alpha. ...
```
Same questions as just above :)

wk8 commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,152 @@ A developer has a Windows web app that they would like to deploy in a container
+
+6. A mutating webhook admission controller, GMSAAuthorizer, will act on pod creations and execute the following:
+
+  - Check the pod spec for references to names of GMSACredSpec objects in the pod (and per-container) annotations [for Alpha] or securityContext fields [Beta onwards]. Make sure the GMSACredSpec objects exist and use the contents of the `credspec` field of the GMSACredSpec to populate annotations `pod.alpha.kubernetes.io/windows-gmsa-credspec` and `container.windows-gmsa-credspec.alpha.kubernetes ...
```
> If the specified names of GMSACredSpec cannot be resolved, pod creation will be failed with 404: Not Found.

The current behaviour in my PR is 400 bad request in that case. Happy to change if 404 seems more relevant, your call :)

ddebroy replied:
Not an expert but I was thinking 404 is slightly more specific (like 403 for authz failures).

wk8 replied:
Fair enough, will change.

ddebroy commented:
@liggitt @yujuhong @PatrickLang I have incorporated feedback from the most recent reviews this week. PTAL.

michmike commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -6,18 +6,20 @@ authors:
-  - "sig-arch"
+  - @liggitt
+  - @yujuhong
+  - @patricklang
```
we should have someone from sig-architecture also be an approver

michmike commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
+1. An Active Directory domain administrator will:
 
   - Join all Windows Kubernetes nodes to the Active Directory domain.
   - Provision the GMSA and gives details to application admin.
```
specifically, it is not the application admin that needs the GMSA details, but the Kubernetes cluster admin. `- Provision the GMSA and gives details to application admin.`

the application admin/developer will only need the artifact that is produced by the kubernetes cluster admin

ddebroy replied:
Agree .. will change.

michmike commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
+  - ActiveDirectory
+  DomainJoinConfig:
+    DnsName: contoso.com
+    DnsTreeName: contoso.com
```
why is our definition of the GMSACredSpec different from the details that active directory defines for a GMSA?
the fields in AD for gmsa as returned by `Get-ADServiceAccount -Identity webapplication1` are
1. distringuished name: cn=webapplication1,CN=managed service accounts,CN=contoso,CN=com
2. name = webapplication1
3. SID = <sid>
There is no netbios name or DNS name (those are inferred from the distinguished name)

ddebroy replied:
The above is based on the output of `Get-CredentialSpec` as Microsoft recommends in: https://raw.githubusercontent.com/Microsoft/Virtualization-Documentation/live/windows-server-container-tools/ServiceAccounts/CredentialSpec.psm1 
and described in 
https://blogs.msdn.microsoft.com/containerstuff/2017/01/30/create-a-container-with-active-directory-support/
I will add refs to the above to clarify. The only difference from what is generated by the above powershell scripts is I am suggesting the data be stored in YAML to align with everything else in Kubernetes and changed to JSON when necessary as a blob.

michmike replied:
@ddebroy  thanks for clarifying. it is interesting that they are different in structure, but i am ok with your approach

michmike replied:
@ddebroy  one last thing. if the scope of the GMSA is more than one node (as defined in active directory), how should that be represented in the yaml? you don't have to put such an example in the KEP, but we should have that in our docs.

michmike commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
+    - If a credspec JSON is not specified already, GMSAAuthorizer will look up the corresponding GMSACredSpec name and use the credspec JSON prepared from the previous step to populate the annotations or securityContext fields for credspec JSON.
+    - If a credspec JSON is specified, look up the corresponding GMSACredSpec name and perform a deep equal check of the members of the specified credspec JSON and prepared credspec JSON from previous step. If the deep equal check fails, GMSAAuthorizer will be fail pod creation with 400: BadRequest.
+
+  - Check the specified ServiceAccount is authorized for the `use` verb on all specified GMSACredSpecs at the pod level and for each container. If the authorization check fails due to absence of requisite RBAC roles, the pod creation will be failed with a 403: Forbidden. 
```
if RBAC is not enabled, i think we should also fail with 403: Forbidden (if that's not already taken care of)

ddebroy replied:
Today, if RBAC and PSP is not enabled, starting a pod with privileged containers or extra capabilities is not failed/forbidden (please correct me if I am wrong). I think we should align with this existing model for gMSA and leave it up to the kubernetes cluster administrator (or distro vendors' setup mechanisms/cluster setup script authors) to configure and enable RBAC and the GMSAAuthorizer webhook if they want policy and authorization around gMSAs and service accounts in their clusters. Thoughts on keeping the core non opinionated in the context of gMSA and policy/authorization?

michmike replied:
I added it as a topic to discuss in our SIG-windows meet up today. i am torn half way on this, but i generally now that i think about this some more agree with being non-opinionated. in our docs we can stress the importance of RBAC in this context

ddebroy replied:
Based on discussion in SIG Windows, the conclusion was to stay non opinionated in this regard i.e. not enforce enablement of RBAC. This also aligns with how privileged pods and similar concepts are allowed without PSP/RBAC.

michmike commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
+      Reason:           CrashLoopBackOff
+Last State:         Terminated
+      Reason:           ContainerCannotRun
+      Message:          [runtime specific error e.g. "encountered an error during CreateContainer:..."]
```
it is not clear to me that we will provide a very specific error message that GMSA is what failed to start the container. we def need to do that.

ddebroy replied:
What I meant above is that the specific error in processing the GMSA is detected and returned to Kubernetes CRI layer by the container runtime. The container runtime can fail to create a container due to several reasons like incorrect Windows base image, etc besides invalid GMSA config. I will clarify this point above.

I think the container runtime specific error message is slightly beyond the scope of the Kubernetes KEP but we should audit all run times that support GMSA is reporting appropriate error strings to orchestrators that hcsshim in Windows is reporting to them.

ddebroy replied:
I have added a note on how init containers maybe used to get around those scenarios where containers get started even though the credspec configuration is incorrect.

michmike commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
+
+`DetermineEffectiveSecurityContext` will be enhanced to populate `CredentialSpec` for Windows containers based on credspec specification for containers and pod. It will need to obtain the credspec contents from the `pod.alpha.kubernetes.io/windows-gmsa-credspec` and `container.windows-gmsa-credspec.alpha.kubernetes.io/container-name` annotations during the Alpha phase. The enhancements will be en ...
 
 #### Changes in Dockershim
```
how would this change with the containerD changes down the pipeline

yujuhong replied:
Yes, please add a brief section about how this will be implemented in containerd.

ddebroy replied:
Done

michmike commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
-9. During any subsequent pod update, any changes to a pod's `securityContext` is blocked (as is the case today) by `ValidatePodUpdate`.
+5. Application admins will deploy app pods that require a GMSA identity along with a Service Account authorized to use the GMSAs. There will be two ways to specify the GMSA credspec details for pods and containers. It is expected that users will typically use the first option as it is more user friendly. The second option is available mainly due to an artifact of the design choices made and descr ...
+
+  - Specify the name of the desired GMSACredSpec object (e.g. `webapp1-credspec`): If an application administrator wants containers to be initialized with a GMSA identity, specifying the names of the desired GMSACredSpec objects is mandatory. In the alpha stage of this feature, the name of the desired GMSACredSpec can be set through an annotation on the pod (applicable to all containers): `pod.al ...
```
can you also add an example of a POD spec leveraging this feature?

ddebroy replied:
Done

yujuhong commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
 
 #### Changes in Dockershim
-The `applyWindowsContainerSecurityContext` function will create a temporary file with a unique name on the host file system with the contents of the credspec. The file path will be used to populate `HostConfig.SecurityOpt` with a credspec file specification. The credspec file will be deleted after `StartContainer` in invoked as well as in various error handling paths between container creation an ...
+The `applyWindowsContainerSecurityContext` function will create a temporary file with a unique name on the host file system with the contents of `WindowsContainerSecurityContext.CredentialSpec`. The file path will be used to populate `HostConfig.SecurityOpt` with a credspec file specification. The credspec file will be deleted after `StartContainer` in invoked as well as in various error handling ...
```
What would be the path to the file?

ddebroy replied:
The file path will be under `C:\ProgramData\docker\CredentialSpecs\`. Will explain in the KEP.

ddebroy commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
+
+  - Check the specified ServiceAccount is authorized for the `use` verb on all specified GMSACredSpecs at the pod level and for each container. If the authorization check fails due to absence of requisite RBAC roles, the pod creation will be failed with a 403: Forbidden. 
+
+7. The Windows CRI implementation will access the credspec JSON through annotations [`container.windows-gmsa-credspec.alpha.kubernetes.io/container-name` for Alpha] or securityContext fields [`podspec.container[i].securityContext.windows.gmsaCredSpec` Beta onwards] for each container in a pod and copy the contents to the [OCI windows.CredentialSpec](https://github.com/opencontainers/runtime-spec/ ...
```
Kubelet.exe will use an enhanced version of `DetermineEffectiveSecurityContext` (that takes into account analysis of the credspec at the pod level and substituting it with a container level spec where specified) to populate/embed the entire credspec JSON for a container in either the annotations [for Alpha] or a new Windows specific securityContext field [for Beta]. The annotations are already passed through CRI today and with the credspec JSON embedded there, no further changes in CRI will be necessary for Alpha. The `WindowsContainerSecurityContext.CredentialSpec` [Beta onwards] will indeed need to be passed through the CRI APIs as you point out. I will clarify this more in this section as well in the implementations notes.

ddebroy commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
+
+A new field `CredentialSpec String` will be added to `WindowsContainerSecurityContext`. 
+
+`DetermineEffectiveSecurityContext` will be enhanced to populate `CredentialSpec` for Windows containers based on credspec specification for containers and pod. It will need to obtain the credspec contents from the `pod.alpha.kubernetes.io/windows-gmsa-credspec` and `container.windows-gmsa-credspec.alpha.kubernetes.io/container-name` annotations during the Alpha phase. The enhancements will be en ...
```
Yes. Will clarify.

ddebroy replied:
Done

ddebroy commented:
@liggitt @yujuhong @michmike thanks so much again for the detailed review and comments. I have addressed all comments so far and expanded/clarified several of the areas especially around errors and the CRI and kubelet changes. PTAL.

michmike commented:
/lgtm

michmike commented:
/hold
Deep you can remove the hold when ready

liggitt commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -88,119 +93,243 @@ To achieve the scale and speed required for containers, Windows uses a group man
+  names:
+    kind: GMSACredSpec
+    plural: gmsacredspecs
+  scope: Cluster
```
do you plan to define the permitted schema for the CRD? if you don't, arbitrary values can be written to GMSACredSpec instances

michmike replied:
that's a great idea. it is only for one value anyway.
would this work?
validation:
    openAPIV3Schema:
      properties:
        spec:
          properties:
            GMSACredSpec:
              type: string
              pattern: '^(\d+|\*)(/\d+)?(\s+(\d+|\*)(/\d+)?){4}$'

ddebroy replied:
I will add the full validation schema here based on the powershell script and the object schema involved:

```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: gmsacredspecs.auth.k8s.io
spec:
  group: auth.k8s.io
  version: v1alpha1
  names:
    kind: GMSACredSpec
    plural: gmsacredspecs
  scope: Cluster
  validation:
    openAPIV3Schema:
      properties:
        credspec:
          description: GMSA Credential Spec
          properties:
            ActiveDirectoryConfig:
              description: Windows Active Directory configuration
              properties:
                GroupManagedServiceAccounts:
                  description: GMSA names and scopes
                  type: array
                  items:
                    Name:
                      description: Name of GMSA
                      type: string
                    Scope:
                      description: DNSRoot or NetBIOS name of GMSA
                      type: string
                CmsPlugins:
                  description: CMS Plugins
                  type: array
                  items:
                    ActiveDirectory
                DomainJoinConfig:
                  description: Configuration to join the domain
                  properties:
                    DnsName:
                      description: Name of AD domain forest
                      type: string
                    DnsTreeName:
                      description: DNS root of AD domain
                      type: string
                    Guid:
                      description: GUID of AD domain
                      type: string
                    MachineAccountName:
                      description: Name of GMSA
                      type: string
                    NetBiosName:
                      description: NetBios name of AD Domain
                      type: string
                    Sid:
                      description: Security identifier of AD Domain
                      type: string
```

liggitt commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -88,119 +93,243 @@ To achieve the scale and speed required for containers, Windows uses a group man
+  group: auth.k8s.io
+  version: v1alpha1
+  names:
+    kind: GMSACredSpec
```
based on the CRI and docker references, would this be better called "GMSACredentialSpec" or "CredentialSpec"?

liggitt commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -88,119 +93,243 @@ To achieve the scale and speed required for containers, Windows uses a group man
+metadata:
+  name: gmsacredspecs.auth.k8s.io
+spec:
+  group: auth.k8s.io
```
I'm not sure what group I'd expect here... possibly something sig-windows-related?

liggitt commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -88,119 +93,243 @@ To achieve the scale and speed required for containers, Windows uses a group man
-7. The Windows OCI implementation validates the credspec and fails to start the container if its invalid or access to the GMSA from the node is denied.
-8. The container starts with the Windows Group Managed Service Account as expected, and can authenticate to the database.
-9. During any subsequent pod update, any changes to a pod's `securityContext` is blocked (as is the case today) by `ValidatePodUpdate`.
+5. Application admins will deploy app pods that require a GMSA identity along with a Service Account authorized to use the GMSAs. There will be two ways to specify the GMSA credspec details for pods and containers. It is expected that users will typically use the first option as it is more user friendly. The second option is available mainly due to an artifact of the design choices made and descr ...
```
> It is expected that users will typically use the first option as it is more user friendly

if the admission plugin allows direct specification, doesn't that allow bypass of the authz check? I'd expect the admission plugin to override any user-specified credspec annotations on pod creation

ddebroy replied:
@liggitt we specifically have a check for credspec specified in expanded form by the user. It is covered in step [6] as part of the actions the webhook performs:

`If a credspec JSON is specified, look up the corresponding GMSACredSpec name and perform a deep equal check of the members of the specified credspec and prepared credspec from previous step. If the deep equal check fails, GMSAAuthorizer will be fail pod creation with 400: BadRequest.`

Note also that we specify that a name referring to a GMSACredSpec is mandatory (regardless of whether the expanded contents is specified so that the name can be validated against): 

`If an application administrator wants containers to be initialized with a GMSA identity, specifying the names of the desired GMSACredSpec objects is mandatory
`

liggitt replied:
ah, hadn't gotten down to that yet. thanks.

liggitt commented on keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md:
```diff
@@ -88,119 +93,243 @@ To achieve the scale and speed required for containers, Windows uses a group man
+  name: webapp-sa
   apiGroup: rbac.authorization.k8s.io
 roleRef:
   kind: Role
```
nit, this is a ClusterRole in the example
